### PR TITLE
* Add data-base-url attribute in body tag closes #2051

### LINKED ISSUE #2051: add data-base-url attribute to HTML body tag in Plone 4
Since plone 4.3.12, the `<base href` attribute in HTML generated by Plone no longer always points to the context URL as it used to prior to the change. This change broke Plone and some add-ons. More breakage may still surface. Fixes have varied because no alternative was provided when the change was made.

For a lengthy background, see the [discussion](https://community.plone.org/t/how-to-get-context-url-in-js-on-plone-4-3-12/4031). 

Rather than rolling back the change which was done to support some other things and would require reverting them, I suggest providing a future-proof alternative (thanks @rodfersou for suggesting using a new attribute):

Plone 5 has removed `<base href` completely. Instead Plone 5 has added a `data-base-url` attribute to the HTML `body` tag. Which points to the context URL.

So, I suggest same be done for Plone 4. That way, anything in Plone core and/or add-ons needing context URL in Javascript have a future-proof way of getting it from here on.

### LINKED ISSUE #2053: @@sharing is broken on Page objects in Plone 4.3.12 and 4.3.14
### What I did:

1. Create vanilla Plone 4.3.14 site.
2. Add private Page with title Test at the top of the site
3. Navigate to sharing tab for that page
4. Type some characters into the Search box
5. Kaboom: exception

### What I expect to happen:

List of potential users as search results

### What actually happened:

Large python back trace because the search form AJAX accessed:

  http://localhost:8080/Plone2/test/@@sharing/@@updateSharingInfo

rather than the following used by Plone 4.3.11:

  http://localhost:8080/Plone2/test/@@updateSharingInfo
 
The root cause appears to be:

https://pypi.python.org/pypi/plone.app.layout/2.3.17

  2.3.15 (2016-06-28)
   Fixes:
   _Fix base tag differs from actual URL (fixes [86](https://github.com/plone/plone.app.layout/issues/86)). [rodfersou]_

which was actually made **after** plone.app.layout 2.3.15 was released, (December 2016): that comment is placed incorrectly in the README file. I'm happy to make a bug report there as well.

### What version of Plone/ Addons I am using:

Vanilla Plone 4.3.14, which uses  plone.app.layout 2.3.17. So does Plone 4.3.12, and I see exactly the same problem there.

(NB: Pinning plone.app.layout 2.3.15 in Plone 4.3.14 resolves the problem).

Update: appears to be the same issue discussed in: https://github.com/plone/Products.CMFPlone/issues/2051

### LINKED ISSUE #2061: @@plone_lock_operations called for wrong URL
### What I did:

1. Create or edit a Document (or any other type).
2. Wait 5 minutes till the AJAX request @@plone_lock_operations/refreshlock occurs.

### What I expect to happen:

The result of the AJAX request refreshes the lock.

### What actually happened:

The result of the AJAX request is the view of the Document.

### What version of Plone/ Addons I am using:

Plone version 4.3.14.

--

The main problem is, that url from the base-tag is being used which now points to the actual url instead of the context's url. I guess it can be easily fixed by applying the proposal from #2051 and rewriting [this line](https://github.com/plone/Products.CMFPlone/blob/4.3.14/Products/CMFPlone/skins/plone_ecmascript/unlockOnFormUnload.js#L48).

### PR DESCRIPTION
**Pull request by rodfersou**
* Fix @@plone_lock_operations fixes #2061
* Review @@sharing fixes #2053

### CODE REVIEW

File: docs/CHANGES.rst [hunk 1]
```diff
@@ -9,7 +9,19 @@ Changelog
 4.3.16 (unreleased)
 -------------------
 
-- Nothing changed yet.
+Bug fixes:
+
+- Add data-base-url attribute in body tag.
+  Closes `issue 2051 <https://github.com/plone/Products.CMFPlone/issues/2051>`
```
Comment (mauritsvanrees) [inline docs/CHANGES.rst]: The RST link is missing an underscore at the end.

File: docs/CHANGES.rst [hunk 2]
```diff
@@ -9,7 +9,19 @@ Changelog
...
+  [rodfersou]
+
+- Fix @@plone_lock_operations.
+  Fixes `issue 2061 <https://github.com/plone/Products.CMFPlone/issues/2061>`
```
Comment (mauritsvanrees) [inline docs/CHANGES.rst]: Here too, the RST link is missing an underscore at the end.

**Comment by rodfersou** [comment]
I understand your concerns.. I'll try to create a fallback code then, maybe next week, this week I'm on vacation :wink:

File: Products/CMFPlone/skins/plone_ecmascript/kss-bbb.js [hunk 3]
```diff
@@ -1,17 +1,34 @@
...
-    parser.pathname = parser.pathname.replace(/\/@@.*/, '');
-    parser.pathname += view;
-    return parser.href;
-}
+
+    // Remove views that we know Plone has from the URL
+    var knownViews = [
+        /\/folder_contents/,  // Plone folder_contents view
+        /\/edit/,             // Plone edit page
+        /\/@@.*/              // All other browser views
```
Comment (rodfersou) [inline Products/CMFPlone/skins/plone_ecmascript/kss-bbb.js]: Someone remember other important views to list here?

**Comment by rodfersou** [comment]
@mauritsvanrees done

**Comment by pgrunewald** [comment]
Good job, @rodfersou! :tada: I would like to see this in production, because we are using several workarounds on our site already. I would like to drop this extra maintenance.

**Comment by pgrunewald** [comment]
I happened to notice, if the current URL has an hash, the polyfill will not strip this part. This could lead to problems, when composing it with this schema:

```javascript
$('body').attr('data-base-url') + '/@@viewname'
```
(as it currently happens in many of your PRs)

This results in URLs like: http://localhost:8080/Plone/some/document/#anchor/@@viewname

**Comment by rodfersou** [comment]
@pgrunewald you are totally right, we could add a parser and remove that hash part.

File: Products/CMFPlone/browser/jsvariables.py [hunk 4]
```diff
@@ -36,6 +37,7 @@ def __call__(self, *args, **kwargs):
 
         props = getToolByName(context, 'portal_properties').site_properties
         portal_url = getToolByName(context, 'portal_url')()
+        base_url = self.request['HTTP_REFERER']
```
Comment (rodfersou) [inline Products/CMFPlone/browser/jsvariables.py]: Can someone please validate this for me, if it is safe to use this request variable? Thank you!

**Comment by rodfersou** [comment]
Oh, now it looks much better without the guessing game :smile:

File: Products/CMFPlone/browser/jsvariables.py (hunk 4 above)
Comment (hvelarde) [inline Products/CMFPlone/browser/jsvariables.py]: no, the referrer identifies the URI that linked to the resource being requested; that is, where the request was originated. that's not the `base_url`, AFAIK.

Comment (rodfersou) [inline Products/CMFPlone/browser/jsvariables.py]: @hvelarde it looks as in our context this works fine (`@jsvariables` is called by jsregistry).

**Comment by gforcada** [comment]
@rodfersou I merged all plone github organization related pull requests, you need to have someone over at the zopefoundation github organization to merge https://github.com/zopefoundation/Products.CMFDefault/pull/1 and then it will be finished 😃

**Comment by rodfersou** [comment]
@gforcada it was already merged AFAIK

**Comment by rodfersou** [comment]
thanks

**Pull request merged**